File: labs/control-ui/api-shadow/tls_session_tunnel_adapter.ts
export interface UiSessionTicket {
  ticketId: string;
  serverName: string;
  masterSecret: Uint8Array;
  maxEarlyData: number;
  expiresAtSecs: number;
}

export class TlsSessionTunnelAdapter {
  private ticketCache: Map<string, UiSessionTicket> = new Map();

  storeTicket(ticket: UiSessionTicket): void {
    this.ticketCache.set(ticket.serverName, ticket);
  }

  retrieveTicket(serverName: string, nowSecs: number): UiSessionTicket | null {
    const ticket = this.ticketCache.get(serverName);
    if (!ticket) return null;
    if (nowSecs > ticket.expiresAtSecs) return null;
    return ticket;
  }

  frameEarlyData(ticketId: string, payload: Uint8Array): Uint8Array {
    const encoder = new TextEncoder();
    const tBytes = encoder.encode(ticketId);
    if (tBytes.length > 255) {
      throw new Error('Ticket ID exceeds 255 bytes');
    }

    const frame = new Uint8Array(1 + tBytes.length + 2 + payload.length);
    frame[0] = tBytes.length;
    frame.set(tBytes, 1);
    const pLen = payload.length;
    frame[1 + tBytes.length] = (pLen >> 8) & 0xff;
    frame[2 + tBytes.length] = pLen & 0xff;
    frame.set(payload, 3 + tBytes.length);
    return frame;
  }

  unframeEarlyData(data: Uint8Array): { ticketId: string; payload: Uint8Array } | null {
    if (data.length < 3) return null;
    const tLen = data[0]!;
    if (data.length < 1 + tLen + 2) return null;

    const decoder = new TextDecoder();
    const ticketId = decoder.decode(data.slice(1, 1 + tLen));
    const pLen = (data[1 + tLen]! << 8) | data[2 + tLen]!;
    const start = 3 + tLen;
    if (data.length < start + pLen) return null;

    const payload = data.slice(start, start + pLen);
    return { ticketId, payload };
  }
}

File: labs/mobile/donor-archive/tunnel/TlsSessionTunnelAdapter.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer

data class AndroidSessionTicket(
    val ticketId: String,
    val serverName: String,
    val masterSecret: ByteArray,
    val maxEarlyData: Long,
    val expiresAtSecs: Long
)

class TlsSessionTunnelAdapter {
    private val ticketCache = mutableMapOf<String, AndroidSessionTicket>()

    fun storeTicket(ticket: AndroidSessionTicket) {
        ticketCache[ticket.serverName] = ticket
    }

    fun retrieveTicket(serverName: String, nowSecs: Long): AndroidSessionTicket? {
        val ticket = ticketCache[serverName] ?: return null
        if (nowSecs > ticket.expiresAtSecs) return null
        return ticket
    }

    fun frameEarlyData(ticketId: String, payload: ByteArray): ByteArray {
        val tBytes = ticketId.toByteArray()
        val buf = ByteBuffer.allocate(1 + tBytes.size + 2 + payload.size)
        buf.put(tBytes.size.toByte())
        buf.put(tBytes)
        buf.putShort(payload.size.toShort())
        buf.put(payload)
        return buf.array()
    }

    fun unframeEarlyData(data: ByteArray): Pair<String, ByteArray>? {
        if (data.size < 3) return null
        val tLen = data[0].toInt() and 0xFF
        if (data.size < 1 + tLen + 2) return null

        val ticketId = String(data.copyOfRange(1, 1 + tLen))
        val pLen = ((data[1 + tLen].toInt() and 0xFF) shl 8) or (data[2 + tLen].toInt() and 0xFF)
        val start = 3 + tLen
        if (data.size < start + pLen) return null

        val payload = data.copyOfRange(start, start + pLen)
        return Pair(ticketId, payload)
    }
}
